util/buildRelease/gen_release

```diff
@@ -178,122 +102,259 @@ if (exists($ENV{"CHPL_GEN_RELEASE_NO_CLONE"})) {
-
-print "[gen_release] Left result in $tarball_name\n";
-
-chdir $origCwd;
+    "tools/chpl-language-server",
+]
+
+
+def main():
+    version = ""
+    if len(sys.argv) > 1:
+        version = sys.argv[1]
+
+    reldir = f"chapel-{version}" if version else "chapel"
+
+    # Find CHPL_HOME. If unset, use root of repo relative to this script.
+    chplhome = os.getenv(
+        "CHPL_HOME",
+        os.path.dirname(
+            os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+        ),
+    )
+
+    # pointers to temporary directory where chapel will be built
+    basetmpdir = os.getenv("CHPL_GEN_RELEASE_TMPDIR", tempfile.gettempdir())
+    user = os.getlogin()
+    tmpdir = tempfile.mkdtemp(
+        prefix=f"chapel-release.{user}.deleteme.", dir=basetmpdir
+    )
+    archive_dir = pjoin(tmpdir, reldir)
+    rootdir = pjoin(tmpdir, "chpl_home")
+
+    rm_rf(tmpdir)
+    os.makedirs(archive_dir, exist_ok=True)
+
+    # TODO: can we use pythons builtin tar module instead of system commands?
+    tar_executable = "gtar" if shutil.which("gtar") else "tar"
+
+    # If CHPL_GEN_RELEASE_NO_CLONE is set in environment, do not clone the repo
+    # Just use whatever is found in $CHPL_HOME
+    if "CHPL_GEN_RELEASE_NO_CLONE" in os.environ:
+        # copy the current CHPL_HOME into the directory where chapel will be built
+        print(
+            "[gen_release] CHPL_GEN_RELEASE_NO_CLONE: Creating build workspace with tar..."
+        )
+        with cd(chplhome):
+            p = run(tar_executable, "-cf", "-", ".", stdout=subprocess.PIPE)
+            with cd(archive_dir):
+                run(tar_executable, "-xf", "-", input=p.stdout)
+            del p # free all the memory used by the tar file
+        resultdir = pjoin(chplhome, "tar")
+    else:
+        # check out a clean copy of the sources into the directory where chapel will be built
+        git_url = os.getenv(
+            "CHPL_HOME_REPOSITORY", "https://github.com/chapel-lang/chapel"
+        )
+        git_branch = os.getenv("CHPL_GEN_RELEASE_BRANCH", "main")
+        repo_cache_path = os.getenv("REPO_CACHE_PATH", "/missing")
+        git_commit = os.getenv("CHPL_GEN_RELEASE_COMMIT", None)
+
+        print(
+            f"[gen_release] Cloning the sources (repo: {git_url} branch: {git_branch})..."
+        )
+        cmd = (
+            [
+                "git",
+                "clone",
+                "--reference-if-able",
+                f"{repo_cache_path}/chapel.git",
+                "--branch",
+                git_branch,
+            ]
+            + (["--depth", "1"] if git_commit is None else [])
+            + [
+                git_url,
+                rootdir,
+            ]
+        )
+        run(*cmd)
+
+        if git_commit is not None:
+            print(f"[gen_release] Checking out revision {git_commit}...")
+            with cd(rootdir):
+                run("git", "reset", "--hard", git_commit)
+
+        with cd(rootdir):
```

Comment:
can you merge these by moving the above `if git_commit` into the second `with cd`?